In [4]:
import pandas as pd
import numpy as np

df_with_debt_to_income_ratio_scaled = pd.read_csv('with_debt_to_income_ratio_scaled.csv')
df_with_debt_to_income_ratio_scaled

Age    Income  Credit_Score  Loan_Amount  Loan_Term  Interest_Rate  \
0     0.835671 -1.448440      1.660539     1.051966  -0.013857       1.381737   
1     1.707517 -1.624213      1.484741    -1.538473  -0.013857      -0.629330   
2     0.165020  1.559323     -0.405084     1.431974   0.983029      -0.594030   
3    -0.773891 -0.762464     -0.141388    -0.871927  -0.013857       1.452842   
4     1.103931 -1.269255     -1.491263     1.390356   0.484586      -1.680294   
...        ...       ...           ...          ...        ...            ...   
9995  0.768606 -1.622943      0.800386    -1.507363  -0.512300      -1.195298   
9996  0.500346 -0.652979      0.348334    -0.139767   0.983029      -1.148017   
9997  0.902736  1.581543      1.139424     0.577711   0.983029       1.034760   
9998  1.372192 -0.028239      1.302665     0.754506  -1.509186       0.709208   
9999 -0.773891  0.608405      1.396842    -0.689899  -0.512300       0.028605   

      Monthly_Installment  Debt_to_Income_Ratio  Employment_Length  \
0               -1.450783              1.277602          -1.269251   
1               -0.710394              0.108876          -1.442360   
2               -0.490734             -1.522750          -1.528915   
3               -0.774257             -1.190992          -1.269251   
4               -1.395341              1.533409          -0.490262   
...                   ...                   ...                ...   
9995            -1.439554              0.783177          -1.442360   
9996            -0.951810             -1.343202           0.029065   
9997             1.519192             -1.507659          -1.269251   
9998             1.667971              0.402935          -1.615469   
9999             1.051098              1.498746           1.413936   

      Savings_Amount  ...  Cholesterol_Level  Exercise_Hours_Per_Week  \
0           0.238627  ...           0.111821                 0.967583   
1           0.713674  ...           0.653918                -0.285916   
2          -0.609242  ...          -1.456388                -1.255940   
3           1.362264  ...          -0.101146                 0.696847   
4           1.583146  ...          -1.553191                -0.807610   
...              ...  ...                ...                      ...   
9995       -0.690364  ...          -1.669355                 1.364845   
9996       -1.044514  ...          -0.894931                -1.410120   
9997       -1.612446  ...          -1.107897                 0.458269   
9998        0.522340  ...           0.015018                 1.447290   
9999        1.501762  ...          -1.746798                 0.320599   

      Caloric_Intake  Sleep_Hours  Social_Media_Usage  \
0           0.339351     1.530153            1.572174   
1           0.433063     1.634319           -1.567708   
2          -0.823782    -1.064924           -0.869957   
3           1.404637     1.461849           -0.521081   
4          -1.523867     0.412949            0.874422   
...              ...          ...                 ...   
9995       -0.659786     0.400612           -0.172205   
9996       -0.141613     0.320558            1.223298   
9997        0.952614     1.004336            1.572174   
9998       -0.560561     0.307000            0.525547   
9999       -0.949191     1.541912            1.572174   

      Streaming_Subscriptions  Gaming_Hours  Work_Hours_Per_Week  \
0                   -0.693788     -0.402073             1.280842   
1                   -0.693788     -0.727396            -0.808761   
2                    1.419057      1.175233             1.529604   
3                   -0.693788     -0.774416             0.584308   
4                   -0.693788     -1.275717            -0.211731   
...                       ...           ...                  ...   
9995                -0.693788      0.751481             0.136536   
9996                 0.010494      0.771365            -1.704305   
9997                -0.

In [5]:
df_without_debt_to_income_ratio_scaled = pd.read_csv('without_debt_to_income_ratio_scaled.csv')
df_without_debt_to_income_ratio_scaled

Age    Income  Credit_Score  Loan_Amount  Loan_Term  Interest_Rate  \
0     0.835671 -1.448440      1.660539     1.051966  -0.013857       1.381737   
1     1.707517 -1.624213      1.484741    -1.538473  -0.013857      -0.629330   
2     0.165020  1.559323     -0.405084     1.431974   0.983029      -0.594030   
3    -0.773891 -0.762464     -0.141388    -0.871927  -0.013857       1.452842   
4     1.103931 -1.269255     -1.491263     1.390356   0.484586      -1.680294   
...        ...       ...           ...          ...        ...            ...   
9995  0.768606 -1.622943      0.800386    -1.507363  -0.512300      -1.195298   
9996  0.500346 -0.652979      0.348334    -0.139767   0.983029      -1.148017   
9997  0.902736  1.581543      1.139424     0.577711   0.983029       1.034760   
9998  1.372192 -0.028239      1.302665     0.754506  -1.509186       0.709208   
9999 -0.773891  0.608405      1.396842    -0.689899  -0.512300       0.028605   

      Monthly_Installment  Employment_Length  Savings_Amount  \
0               -1.450783          -1.269251        0.238627   
1               -0.710394          -1.442360        0.713674   
2               -0.490734          -1.528915       -0.609242   
3               -0.774257          -1.269251        1.362264   
4               -1.395341          -0.490262        1.583146   
...                   ...                ...             ...   
9995            -1.439554          -1.442360       -0.690364   
9996            -0.951810           0.029065       -1.044514   
9997             1.519192          -1.269251       -1.612446   
9998             1.667971          -1.615469        0.522340   
9999             1.051098           1.413936        1.501762   

      Stock_Investment  ...  Cholesterol_Level  Exercise_Hours_Per_Week  \
0             0.033889  ...           0.111821                 0.967583   
1            -0.976664  ...           0.653918                -0.285916   
2            -0.099320  ...          -1.456388                -1.255940   
3            -1.072342  ...          -0.101146                 0.696847   
4             0.543112  ...          -1.553191                -0.807610   
...                ...  ...                ...                      ...   
9995         -0.861710  ...          -1.669355                 1.364845   
9996          0.194444  ...          -0.894931                -1.410120   
9997         -0.986390  ...          -1.107897                 0.458269   
9998         -1.205515  ...           0.015018                 1.447290   
9999         -1.626708  ...          -1.746798                 0.320599   

      Caloric_Intake  Sleep_Hours  Social_Media_Usage  \
0           0.339351     1.530153            1.572174   
1           0.433063     1.634319           -1.567708   
2          -0.823782    -1.064924           -0.869957   
3           1.404637     1.461849           -0.521081   
4          -1.523867     0.412949            0.874422   
...              ...          ...                 ...   
9995       -0.659786     0.400612           -0.172205   
9996       -0.141613     0.320558            1.223298   
9997        0.952614     1.004336            1.572174   
9998       -0.560561     0.307000            0.525547   
9999       -0.949191     1.541912            1.572174   

      Streaming_Subscriptions  Gaming_Hours  Work_Hours_Per_Week  \
0                   -0.693788     -0.402073             1.280842   
1                   -0.693788     -0.727396            -0.808761   
2                    1.419057      1.175233             1.529604   
3                   -0.693788     -0.774416             0.584308   
4                   -0.693788     -1.275717            -0.211731   
...                       ...           ...                  ...   
9995                -0.693788      0.751481             0.136536   
9996                 0.010494      0.771365            -1.704305   
9997                -0.693788     -1.150489             0.086783   
999

In [6]:
df_with_debt_to_income_ratio_scaled_loan_status = df_with_debt_to_income_ratio_scaled.drop(columns=['Churn_Status'])
df_without_debt_to_income_ratio_scaled_loan_status = df_without_debt_to_income_ratio_scaled.drop(columns=['Churn_Status'])



df_with_debt_to_income_ratio_scaled_churn_status = df_with_debt_to_income_ratio_scaled.drop(columns=['Default_Status'])
df_without_debt_to_income_ratio_scaled_churn_status = df_without_debt_to_income_ratio_scaled.drop(columns=['Default_Status'])

In [7]:
df_with_debt_to_income_ratio_scaled_loan_status_X = df_with_debt_to_income_ratio_scaled_loan_status.drop(columns=['Default_Status'])
df_with_debt_to_income_ratio_scaled_loan_status_Y = df_with_debt_to_income_ratio_scaled_loan_status['Default_Status']


df_without_debt_to_income_ratio_scaled_loan_status_X = df_without_debt_to_income_ratio_scaled_loan_status.drop(columns=['Default_Status'])
df_without_debt_to_income_ratio_scaled_loan_status_Y = df_without_debt_to_income_ratio_scaled_loan_status['Default_Status']





df_with_debt_to_income_ratio_scaled_churn_status_X = df_with_debt_to_income_ratio_scaled_churn_status.drop(columns=['Churn_Status'])
df_with_debt_to_income_ratio_scaled_churn_status_Y = df_with_debt_to_income_ratio_scaled_churn_status['Churn_Status']

df_without_debt_to_income_ratio_scaled_churn_status_X = df_without_debt_to_income_ratio_scaled_churn_status.drop(columns=['Churn_Status'])
df_without_debt_to_income_ratio_scaled_churn_status_Y = df_without_debt_to_income_ratio_scaled_churn_status['Churn_Status']


## Prediction for if the person will get loan

In [8]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# Split into train and test sets (80% train, 20% test)
X_train_loan1, X_test_loan1, y_train_loan1, y_test_loan1  = train_test_split(df_with_debt_to_income_ratio_scaled_loan_status_X, df_with_debt_to_income_ratio_scaled_loan_status_Y, test_size=0.2, random_state=42)


X_train_loan11, X_test_loan11, y_train_loan11, y_test_loan11  = train_test_split(df_without_debt_to_income_ratio_scaled_loan_status_X, df_without_debt_to_income_ratio_scaled_loan_status_Y, test_size=0.2, random_state=42)

## Predicting if a customer will churn

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# Split into train and test sets (80% train, 20% test)
X_train_churn2, X_test_churn2, y_train_churn2, y_test_churn2  = train_test_split(df_with_debt_to_income_ratio_scaled_churn_status_X, df_with_debt_to_income_ratio_scaled_churn_status_Y, test_size=0.2, random_state=42)


X_train_churn22, X_test_churn22, y_train_churn22, y_test_churn22  = train_test_split(df_without_debt_to_income_ratio_scaled_churn_status_X, df_without_debt_to_income_ratio_scaled_churn_status_Y, test_size=0.2, random_state=42)

In [21]:
model1 = LogisticRegression()
model11 = LogisticRegression()
model2 = LogisticRegression()
model22 = LogisticRegression()

In [22]:
model1.fit(X_train_loan1, y_train_loan1)
model11.fit(X_train_loan11, y_train_loan11)
model2.fit(X_train_churn2, y_train_churn2)
model22.fit(X_train_churn22, y_train_churn22)

LogisticRegression()

In [23]:
# Predict on test data
y_pred1 = model1.predict(X_test_loan1)
y_pred11 = model11.predict(X_test_loan11)
y_pred2 = model2.predict(X_test_churn2)
y_pred22 = model22.predict(X_test_churn22)

In [24]:
# Evaluate Model
accuracy = accuracy_score(y_test_loan1, y_pred1)
print(f'Accuracy: {accuracy:.2f}')
print('Classification Report:')
print(classification_report(y_test_loan1, y_pred1))

Accuracy: 0.93
Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.96      0.96      1577
           1       0.84      0.83      0.83       423

    accuracy                           0.93      2000
   macro avg       0.90      0.89      0.89      2000
weighted avg       0.93      0.93      0.93      2000



In [25]:
# Evaluate Model
accuracy = accuracy_score(y_test_loan11, y_pred11)
print(f'Accuracy: {accuracy:.2f}')
print('Classification Report:')
print(classification_report(y_test_loan11, y_pred11))

Accuracy: 0.76
Classification Report:
              precision    recall  f1-score   support

           0       0.80      0.93      0.86      1577
           1       0.33      0.13      0.19       423

    accuracy                           0.76      2000
   macro avg       0.57      0.53      0.53      2000
weighted avg       0.70      0.76      0.72      2000



In [26]:
# Evaluate Model
accuracy = accuracy_score(y_test_churn2, y_pred2)
print(f'Accuracy: {accuracy:.2f}')
print('Classification Report:')
print(classification_report(y_test_churn2, y_pred2))

Accuracy: 0.99
Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      1953
           1       0.88      0.81      0.84        47

    accuracy                           0.99      2000
   macro avg       0.94      0.90      0.92      2000
weighted avg       0.99      0.99      0.99      2000



In [27]:
# Evaluate Model
accuracy = accuracy_score(y_test_churn22, y_pred22)
print(f'Accuracy: {accuracy:.2f}')
print('Classification Report:')
print(classification_report(y_test_churn22, y_pred22))

Accuracy: 0.99
Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      1953
           1       0.91      0.83      0.87        47

    accuracy                           0.99      2000
   macro avg       0.95      0.91      0.93      2000
weighted avg       0.99      0.99      0.99      2000



## Predict on Training Data (Overfitting Check)

In [29]:
y_train_pred = model1.predict(X_train_loan1)
train_accuracy = accuracy_score(y_train_loan1, y_train_pred)
print(f"Training Accuracy: {train_accuracy:.2f}")

Training Accuracy: 0.93


In [30]:
y_train_pred_ = model11.predict(X_train_loan11)
train_accuracy_ = accuracy_score(y_train_loan11, y_train_pred_)
print(f"Training Accuracy: {train_accuracy_:.2f}")

Training Accuracy: 0.78


## Predict on Test Data (Generalization Check)

In [31]:
y_test_pred = model1.predict(X_test_loan1)
test_accuracy = accuracy_score(y_test_loan1, y_test_pred)
print(f"Test Accuracy: {test_accuracy:.2f}")

Test Accuracy: 0.93


In [32]:
y_test_pred__ = model11.predict(X_test_loan11)
test_accuracy__ = accuracy_score(y_test_loan11, y_test_pred__)
print(f"Test Accuracy: {test_accuracy__:.2f}")

Test Accuracy: 0.76


## Predict on Training Data (Overfitting Check)

In [34]:
# X_train_churn2, X_test_churn2, y_train_churn2, y_test_churn2
y_train_pred2 = model2.predict(X_train_churn2)
train_accuracy2 = accuracy_score(y_train_churn2, y_train_pred2)
print(f"Training Accuracy: {train_accuracy:.2f}")

Training Accuracy: 0.99


In [35]:
y_train_pred22 = model22.predict(X_train_churn22)
train_accuracy22 = accuracy_score(y_train_churn22, y_train_pred22)
print(f"Training Accuracy: {train_accuracy22:.2f}")

Training Accuracy: 0.99


## Predict on Test Data (Generalization Check)

In [36]:
y_test_pred5 = model2.predict(X_test_churn2)
test_accuracy5 = accuracy_score(y_test_churn2, y_test_pred5)
print(f"Test Accuracy: {test_accuracy5:.2f}")

Test Accuracy: 0.99


In [37]:
y_test_pred51 = model22.predict(X_test_churn22)
test_accuracy51 = accuracy_score(y_test_churn22, y_test_pred51)
print(f"Test Accuracy: {test_accuracy51:.2f}")

Test Accuracy: 0.99


## Checking overfitting for churning status

In [39]:
# Count occurrences of each class
print(df_without_debt_to_income_ratio_scaled_churn_status['Churn_Status'].value_counts(normalize=True))

Churn_Status
0    0.9776
1    0.0224
Name: proportion, dtype: float64


In [40]:
print(df_with_debt_to_income_ratio_scaled_churn_status['Churn_Status'].value_counts(normalize=True))

Churn_Status
0    0.9776
1    0.0224
Name: proportion, dtype: float64


## This is highly imbalaned dataset with no churn more than churn so overfitting checking recall

In [41]:
from sklearn.metrics import classification_report

print(classification_report(y_test_churn2, y_test_pred5))


              precision    recall  f1-score   support

           0       1.00      1.00      1.00      1953
           1       0.88      0.81      0.84        47

    accuracy                           0.99      2000
   macro avg       0.94      0.90      0.92      2000
weighted avg       0.99      0.99      0.99      2000



In [42]:
from sklearn.metrics import classification_report

print(classification_report(y_test_churn22, y_test_pred51))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00      1953
           1       0.91      0.83      0.87        47

    accuracy                           0.99      2000
   macro avg       0.95      0.91      0.93      2000
weighted avg       0.99      0.99      0.99      2000



## Improve recall for people who will churn- as capturing no churn (majority class) more try to oversample churn (minority class)

In [45]:
from imblearn.over_sampling import SMOTE
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
# X_train_churn22, X_test_churn22, y_train_churn22, y_test_churn22
# Apply SMOTE with a higher oversampling ratio (70% of the majority class)
smote = SMOTE(sampling_strategy=0.7, random_state=42)
X_train_churn22_resampled, y_train_churn22_resampled = smote.fit_resample(X_train_churn22, y_train_churn22)

In [46]:
# Train logistic regression with class weighting
model_resampled = LogisticRegression(class_weight={0: 1, 1: 3}, random_state=42)
model.fit(X_train_churn22_resampled, y_train_churn22_resampled)

LogisticRegression()

In [50]:
# Predict probabilities
y_pred_proba = model.predict_proba(X_test_churn22)[:, 1]

# Adjust decision threshold from 0.5 to 0.4 for higher recall
y_pred_adjusted = (y_pred_proba >  0.35).astype(int)

# Evaluate the model
report = classification_report(y_test_churn22, y_pred_adjusted, output_dict=True)

In [51]:
df_report = pd.DataFrame(report).transpose()
df_report

precision    recall  f1-score    support
0              0.999479  0.981567  0.990442  1953.0000
1              0.560976  0.978723  0.713178    47.0000
accuracy       0.981500  0.981500  0.981500     0.9815
macro avg      0.780227  0.980145  0.851810  2000.0000
weighted avg   0.989174  0.981500  0.983926  2000.0000

#### High Recall for Churn (1) (97.87%) means we are correctly identifying most churned customers.
#### Lower Precision for Churn (1) (59.74%) means some non-churn customers are mistakenly classified as churn.

## Trying Random Forest Classifier

In [63]:
from sklearn.ensemble import RandomForestClassifier
# Train a Random Forest Classifier with class weights
rf_model = RandomForestClassifier(n_estimators=100, max_depth=5, class_weight={0: 1, 1: 1.5}, random_state=42)
rf_model.fit(X_train_churn22_resampled, y_train_churn22_resampled)

# Predict probabilities using Random Forest
y_pred_proba_rf = rf_model.predict_proba(X_test_churn22)[:, 1]

# Adjust decision threshold (trying 0.35)
y_pred_adjusted_rf = (y_pred_proba_rf > 0.40).astype(int)

# Evaluate the model
rf_report = classification_report(y_test_churn22, y_pred_adjusted_rf, output_dict=True)

# Display classification report
df_rf_report = pd.DataFrame(rf_report).transpose()
df_rf_report

precision    recall  f1-score   support
0              1.000000  0.995904  0.997948  1953.000
1              0.854545  1.000000  0.921569    47.000
accuracy       0.996000  0.996000  0.996000     0.996
macro avg      0.927273  0.997952  0.959758  2000.000
weighted avg   0.996582  0.996000  0.996153  2000.000